lcls_naming_tool/lcls_naming_tool.py

```diff
@@ -0,0 +1,309 @@
+#!/usr/bin/env python3
+
+'''
+    LCLS Naming Tool
+
+    A tool for checking the form and content (but not meaning) of PV names with
+    respect to the LCLS naming convention.
+
+    Functional component source letters and beampath numbers come from 
+    https://docs.google.com/drawings/d/1Sg7pT2TlKq8PgM7UstmJJBLKeqROQeDWbkD0sDoTDdY/edit
+
+    A valid PV name is in the format FFFFF:GGG:CCC:NN:XXXX, where GGG and NN are
+    optional and can be omitted. The XXXX is omitted for device names.
+
+    All permutations of valid PV and device names according to 
+    LCLS_Photon_Source_and_Systems_Nomenclature.pdf. Names ending in XXXX are 
+    PV names. Without XXXX are device names.
+
+        FFFFF:CCC
+        FFFFF:CCC:NN
+        FFFFF:CCC:XXXX
+        FFFFF:CCC:NN:XXXX
+        FFFFF:GGG:CCC
+        FFFFF:GGG:CCC:NN
+        FFFFF:GGG:CCC:XXXX
+        FFFFF:GGG:CCC:NN:XXXX
+
+    This script only checks the validity of PV and device names.
+'''
+
+import argparse
+import subprocess
+import sys
+import json
+import re
+
+
+fc_dict = {}
+fg_dict = {}
+ccc_dict = {}
+beam_sources = {'K': '', 'L': ''}
+beam_numbers = {'0': '', '1': '', '2': '', '3': '', '4': '', '5': ''}
+
+
+def get_version():
+    result = subprocess.run(
+        ['git', 'describe', '--tags'], 
+        capture_output=True, 
+        text=True
+    )
+    return (str(result.stdout).strip().split('-'))[0]
+
+def display_version():
+    version = get_version()
+
+    parser = argparse.ArgumentParser(description="Current version of LCLS Naming Tool")
+    parser.add_argument('-v', '--version', action='store_true')
+    args = parser.parse_args()
+    
+    if args.version:
+        print(version)
+
+
+def load_taxons():
+    global fc_dict
+    global fg_dict
+    global ccc_dict
+
+    # Read the json files containing all the taxons
+    with open('taxons/functional_component_taxon.json') as fc_file:
+        fc_dict = json.load(fc_file)
+
+    with open('taxons/fungible_element_taxon.json') as fg_file:
+        fg_dict = json.load(fg_file)
+
+    with open('taxons/ccc_taxon.json') as ccc_file:
+        ccc_dict = json.load(ccc_file)
+
+
+def functional_component_is_valid(fc_taxon):
+
+    fc_length = len(fc_taxon)
+    
+    if (4 < len(fc_taxon) < 9):
```

Comment:
Side question from me not about the code: I didn't realize the fc_taxon could be 8 characters long. I'm curious if you have an example for my learning purposes.

Reply:
Yep, it's mostly to accommodate for the sequence number. The valid component prefixes are all 2 letters long. The source letter is either K or L (for LCLS). The beam path numbers range from 0-5 (taken from the beamline map). In the pdf it looks like there is no limit specified for the sequence number of each component. So, I set a limit of 4 digits for the sequence number just in case. It seems unlikely that a sequence number will ever be greater than 9999. I can lower it to 3 digits if that makes more sense.

(From section 8.1 in LCLS_Photon_Source_and_Systems_Nomenclature.pdf) "The functional component name is formed by prepending a two-letter component mnemonic prefix and sequence number to the beam path designation from Section 7.2. Table 3 shows the complete list of functional photon components and their associated prefixes. Beamline components can be identified as having a two-character prefix based on their function and not a specific design or vendor.  Following the component prefix, a sequence number is assigned to each component increasing monotonically along the beam path beginning with one for the initial appearance of each specific component. When a component has more than nine elements the increment will require two digits (starting with 10) and thus the overall Functional Component Element will have six characters. This would extend to three digits in the unlikely event that there were 99 components. Keep in mind that zero padding is not used in this case. Use 1 and 10, not 01 and 010."

Reply:
Thank you, that makes a lot of sense! Woe be us if we end up with IM3234K0